SU2_CFD/src/solvers/CSolver.cpp

```diff
@@ -3094,6 +3115,182 @@ void CSolver::Read_SU2_Restart_Binary(CGeometry *geometry, const CConfig *config
+  const int src = (rank-1+size) % size; // receive from prev.
+  const int count = sendBuf.size();
+
+  for (int iStep = 0; iStep < size; ++iStep) {
+
+    swap(sendBuf, donorVars);
+
+    if (iStep) {
+      /*--- Odd ranks send and then receive, and vice versa. ---*/
+      if (rank%2) SU2_MPI::Send(sendBuf.data(), count, MPI_DOUBLE, dst, 0, SU2_MPI::GetComm());
+      else SU2_MPI::Recv(donorVars.data(), count, MPI_DOUBLE, src, 0, SU2_MPI::GetComm(), MPI_STATUS_IGNORE);
+
+      if (rank%2==0) SU2_MPI::Send(sendBuf.data(), count, MPI_DOUBLE, dst, 0, SU2_MPI::GetComm());
+      else SU2_MPI::Recv(donorVars.data(), count, MPI_DOUBLE, src, 0, SU2_MPI::GetComm(), MPI_STATUS_IGNORE);
+    }
+
+    /*--- Find the closest target for each donor. ---*/
+
+    vector<su2double> targetDist(nPointDonorMax);
+    vector<unsigned long> iTarget(nPointDonorMax);
+
+    for (auto iDonor = 0ul; iDonor < donorVars.rows(); ++iDonor) {
+      int r=0;
+      adt.DetermineNearestNode(donorVars[iDonor], targetDist[iDonor], iTarget[iDonor], r);
+    }
+
+    /*--- Keep the closest donor for each target (this is separate for OpenMP). ---*/
+
+    for (auto iDonor = 0ul; iDonor < donorVars.rows(); ++iDonor) {
+      const auto iPoint = iTarget[iDonor];
+      const auto dist = targetDist[iDonor];
+
+      if (dist < donorDist[iPoint]) {
+        donorDist[iPoint] = dist;
+        isMapped[iPoint] = true;
+        for (auto iVar = 0ul; iVar < donorVars.cols(); ++iVar)
+          localVars(iPoint,iVar) = donorVars(iDonor,iVar);
+      }
+    }
+  }
+  } // everything goes out of scope except "localVars" and "isMapped"
+
+  /*--- Recursively diffuse the nearest neighbor data. ---*/
+
+  auto nDonor = isMapped;
+
+  for (bool done=false; !done; ++nRecurse) {
+    for (auto iPoint = 0ul; iPoint < nPointDomain; ++iPoint) {
+      /*--- Do not change points that are already interpolated. ---*/
+      if (isMapped[iPoint]) continue;
+
+      /*--- Boundaries to boundaries and domain to domain. ---*/
+      const bool boundary_i = geometry->nodes->GetSolidBoundary(iPoint);
+
+      for (const auto jPoint : geometry->nodes->GetPoints(iPoint)) {
+        if (!isMapped[jPoint]) continue;
+        if (boundary_i != geometry->nodes->GetSolidBoundary(jPoint)) continue;
+
+        nDonor[iPoint]++;
+
+        for (auto iVar = 0ul; iVar < localVars.cols(); ++iVar)
+          localVars(iPoint,iVar) += localVars(jPoint,iVar);
+      }
+
+      if (nDonor[iPoint] > 0) {
+        for (auto iVar = 0ul; iVar < localVars.cols(); ++iVar)
+          localVars(iPoint,iVar) /= nDonor[iPoint];
+        nDonor[iPoint] = true;
+      }
+    }
+
+    /*--- Repeat while all points are not mapped. ---*/
+    done = true;
+    for (auto iPoint = 0ul; iPoint < nPointDomain; ++iPoint) {
+      isMapped[iPoint] = nDonor[iPoint];
+      done &= nDonor[iPoint];
+    }
+  }
+  } // everything goes out of scope except "localVars"
+
+  /*--- Move to Restart_Data in "funny" order. ---*/
```

Comment:
... my favourite order. 
Well, I honestly did not attempt to understand this implementation 🙈 but I would be interested to why 'funny' (I guess the access pattern into localVars is rather random? but i have no clue tbh)

Reply:
I guess formally what that does is sort by global index, which is how data is stored in the files. I'll update that comment.